storage/src/local_file_storage.rs

```diff
@@ -0,0 +1,191 @@
+    it under the terms of the GNU Affero General Public License as
+    published by the Free Software Foundation, either version 3 of the
+    License, or (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Affero General Public License for more details.
+
+    You should have received a copy of the GNU Affero General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+use crate::{PutPayload, Storage, StorageErrorKind, StorageFactory, StorageResult};
+use async_trait::async_trait;
+use std::fmt;
+use std::io::{ErrorKind, SeekFrom};
+use std::ops::Range;
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+use tokio::fs;
+use tokio::io::{AsyncReadExt, AsyncSeekExt};
+
+/// File system compatible storage implementation.
+#[derive(Clone)]
+pub struct LocalFileStorage {
+    root: PathBuf,
+}
+
+impl fmt::Debug for LocalFileStorage {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "LocalFileStorage(root={:?})", &self.root)
+    }
+}
+
+impl LocalFileStorage {
+    /// Creates a file storage instance given a root path.
+    pub fn new(root: PathBuf) -> Self {
+        Self { root }
+    }
+
+    /// Creates a file storage instance given a uri
+    pub fn from_uri(uri: &str) -> StorageResult<LocalFileStorage> {
+        let root_path = uri.split("://").nth(1).ok_or_else(|| {
+            StorageErrorKind::DoesNotExist.with_error(anyhow::anyhow!("Invalid root path: {}", uri))
+        })?;
+        Ok(LocalFileStorage::new(PathBuf::from(root_path)))
+    }
+
+    fn uri(&self, relative_path: &Path) -> String {
+        let full_path = self.root.join(relative_path);
+        format!("file://{}", full_path.to_string_lossy().to_string())
+    }
+}
+
+#[async_trait]
+impl Storage for LocalFileStorage {
+    async fn put(&self, path: &Path, payload: PutPayload) -> crate::StorageResult<()> {
+        let full_path = self.root.join(path);
+        match payload {
+            PutPayload::InMemory(data) => {
+                fs::write(full_path, data).await?;
+            }
+            PutPayload::LocalFile(filepath) => {
+                fs::copy(filepath, full_path).await?;
+            }
+        };
+        Ok(())
+    }
+
+    async fn copy_to_file(&self, path: &Path, output_path: &Path) -> StorageResult<()> {
+        let full_path = self.root.join(path);
+        fs::copy(full_path, output_path).await?;
+        Ok(())
+    }
+
+    async fn get_slice(&self, path: &Path, range: Range<usize>) -> StorageResult<Vec<u8>> {
+        let full_path = self.root.join(path);
+        let mut file = fs::File::open(full_path).await.map_err(|_| {
+            StorageErrorKind::DoesNotExist.with_error(anyhow::anyhow!(
```

Comment:
can you implement a From<io::Error> for storage error that has specific logic for the DoesNotExist case based on the io::ErrorKind.